changelog/7133.improvement.rst

```diff
@@ -0,0 +1 @@
+``caplog.set_level()`` will now override any :confval:`log_level` set via the CLI or ``.ini``.
```

Comment:
Do you agree that tests must be written in a way that always succeed no matter what logging plugin settings are used on the cmd line or in `pytest.ini`? E.g. if a test passes when the user passed `--log-level=INFO` to `pytest` but fails if `--log-level=DEBUG` is used, then sth is wrong with the test.

In the current version of pytest writing tests that are independent of the logging settings (pytest.ini / cmd line) is not possible, but I plan to change that.

Reply:
> Do you agree that tests must be written in a way that always succeed no matter what logging plugin settings are used on the cmd line or in `pytest.ini`?

Yes, very much so, hence the need for this PR in the first place 😉 We need to override what `caplog` sees at the test level such that no matter the CLI or `.ini`, the test still passes.

Reply:
FYI I have a WIP branch, where I try to get rid of the setLevel code for the root logger:
https://github.com/pytest-dev/pytest/compare/master...thisch:root_logger_refactoring